Create Unit Tests for CpdBootStrap

casterkay opened the pull request:
Closes #29

j-emberton committed:
Create test_cpdBootstrapUStarTh4Season20100901

j-emberton committed:
Added chatGPT generated tests

casterkay committed:
convert mock_data to a fixture

casterkay committed:
convert mock_data to a fixture

j-emberton committed:
added first test artifacts

j-emberton committed:
Merge branch '29-generate-unit-tests-for-cpdbootstrap' of github.com:Cambridge-ICCS/ONEFlux into 29-generate-unit-tests-for-cpdbootstrap

casterkay commented on tests/test_artifacts/cpdBootstrapUStarTh4Season20100901_artifacts/cpdBootstrapUStarTh4Season20100901_artifacts.json:
```diff
@@ -0,0 +1,25 @@
+      "id": "Testcase_1",
+      "description": "Basic functionality test using US_ARC_sample reference data",
+      "input": {
+        "0": "t.csv",
```
It seems that `t.csv` is not in the `Testcase_1` folder. Is it because it has name conflict with `T.csv`? Maybe Github does not distinguish between lower and upper cases.

dorchard replied:
I think this should be avoided as some filesystems won't distinguish uppercase from lowercase

j-emberton committed:
first pass at new function etst based on testcases

casterkay committed:
Add test_cpdEvaluateUStarTh4Season and test_cpdFindChangePoint

casterkay committed:
add t.csv

dorchard commented:
For those functions which return an array of structs, we will add an extra optional argument in the MATLAB which flags whether we are testing and do the conversion (either to a struct of arrays or using `struct2cell` or using JSON).

casterkay committed:
fix bugs including unsupported nonscalar struct output, invalid paths of test artifacts, and read_file function

casterkay committed:
Fix bugs: read_csv, CpdBootstrap testcases, cpd values

casterkay commented:
> For those functions which return an array of structs, we will add an extra optional argument in the MATLAB which flags whether we are testing and do the conversion (either to a struct of arrays or using `struct2cell` or using JSON).

I've adopted the JSON encoding approach now.

j-emberton commented on oneflux_steps/ustar_cp/cpdBootstrapUStarTh4Season20100901.m:
```diff
@@ -1,6 +1,6 @@
 	function [Cp2,Stats2,Cp3,Stats3] = ... 
 		cpdBootstrapUStarTh4Season20100901 ...
-			(t,NEE,uStar,T,fNight,fPlot,cSiteYr,nBoot) 
+			(t,NEE,uStar,T,fNight,fPlot,cSiteYr,nBoot,varargin) 
```
This change is just applied to the old reference code. We also have the parallel development version `oneflux_steps/ustar_cp_refactor_wip`, which is where we are refactoring and modularising the code.

I don't mind the idea of adding this here, but the actual code being run by default by the tests cases is the code in the refactor version.

If you look in tests/conftest.py, you'll see the `matlab_engine` fixture, which can be configured to run either the original code or our new code. By default it runs our refactored code. This is why, later in your test you have to manually add the path for this particular function to get the test to run.

So while I think you're doing the right stuff, it'd be good to do it on the refactor codebase as well, and run the tests on that version of the code.

We should try and keep the original as untouched as possible as that is our reference point.

casterkay replied:
I've migrated the changes to `oneflux_steps/ustar_cp_refactor_wip`.

j-emberton commented on oneflux_steps/ustar_cp/cpdEvaluateUStarTh4Season20100901.m:
```diff
@@ -1,6 +1,6 @@
 	function [Cp2,Stats2,Cp3,Stats3] = ... 
 		cpdEvaluateUStarTh4Season20100901 ...
-		(t,NEE,uStar,T,fNight,fPlot,cSiteYr) 
+		(t,NEE,uStar,T,fNight,fPlot,cSiteYr,varargin) 
```
Similar comment here as for the bootstrap function.

casterkay replied:
Resolved as above

j-emberton commented on tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py:
```diff
@@ -0,0 +1,263 @@
+def matlab_engine():
+    # Initialize MATLAB engine
+    eng = matlab.engine.start_matlab()
+    eng.addpath('/home/tcai/Projects/ONEFlux/oneflux_steps/ustar_cp')
```
Adding the path isn't necessary and it's hard coded to your personal machine. You can set the path by overriding the default arg in the matlab_engine picture defined in contest.py

casterkay replied:
Resolved

j-emberton commented on tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py:
```diff
@@ -0,0 +1,263 @@
+def mock_data(nt=300, tspan=(0, 1), uStar_pars=(0.1, 3.5), T_pars=(-10, 30), fNight=None):
+    """
+    Fixture to generate mock time series data for testing purposes. This fixture 
+    creates a set of synthetic data typically used in environmental studies, 
```
```suggestion
    creates a set of synthetic data corresponding to cpdBootstap* function arguments, 
```

j-emberton commented on tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py:
```diff
@@ -0,0 +1,263 @@
+    """
+    Fixture to generate mock time series data for testing purposes. This fixture 
+    creates a set of synthetic data typically used in environmental studies, 
+    such as Net Ecosystem Exchange (NEE), uStar values, temperature, and day/night flags.
```
This line could be more specific.

j-emberton commented on tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py:
```diff
@@ -0,0 +1,263 @@
+    for s2, s3 in zip(Stats2, Stats3):
+        for i in range(8):  # Assuming nStrataX = 8
+            for j in range(nBoot):
+                assert 'n' in s2[i][j]
```
we could add a 4th loop here along the lines of:
for var in ['n', 'CP','fMax','p', 'b0' etc]:
    assert var in s2[I][j]
    assert var in s3[I][j]
    
Which would tidy up this long set of asserts

casterkay replied:
Fixed

j-emberton commented on tests/unit_tests/test_ustar_cp/test_cpdFindChangePoint20100901.py:
```diff
@@ -0,0 +1,57 @@
+def matlab_engine():
+    # Initialize MATLAB engine
+    eng = matlab.engine.start_matlab()
+    eng.addpath('/home/tcai/Projects/ONEFlux/oneflux_steps/ustar_cp')
```
Same thing as before for the hardcoded path and matlab_engine

casterkay replied:
Resolved

casterkay committed:
remove matlab_engine and migrate changes from ustar_cp to ustar_cp_refactor_wip

j-emberton committed:
fixes

j-emberton committed:
moved fixtures and functions into conftest

j-emberton committed:
tidied up tests

j-emberton commented:
I've tidied up the test I wrote and moved a bunch of stuff into test utility functions in `conftest.py`

@tztsai Could you review the suggestions I made previously and resolve.

@dorchard , @isaacaka could you guys review please :)

dorchard commented:
LGTM once those suggestions are resolved by @tztsai

casterkay committed:
Update tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py

Co-authored-by: James Emberton <60827102+j-emberton@users.noreply.github.com>

casterkay commented:
> I've tidied up the test I wrote and moved a bunch of stuff into test utility functions in `conftest.py`
> 
> @tztsai Could you review the suggestions I made previously and resolve.
> 
> @dorchard , @isaacaka could you guys review please :)

I've addressed these comments already in previous commits. I've resolved the discussions now.
